### PR TITLE
SQLite3 Connection getIndexData()

### PR DESCRIPTION
**Pull request by sclubricants**
This is a rewrite of `system/Database/SQLite3/Connection::_indexData()`.

SQLite doesn't show indexes on tables created with `INTEGER PRIMARY KEY AUTO_INCREMENT`. This PR looks for primary keys that don't appear in the normal index lookup and adds it to the return of the method. 

In the test database creation of the users table the current implementation won't return any index.

```php
        $this->forge->addField([
            'id'         => ['type' => 'INTEGER', 'constraint' => 3, 'auto_increment' => true],
            'name'       => ['type' => 'VARCHAR', 'constraint' => 80],
            'email'      => ['type' => 'VARCHAR', 'constraint' => 100],
            'country'    => ['type' => 'VARCHAR', 'constraint' => 40],
            'created_at' => ['type' => 'DATETIME', 'null' => true],
            'updated_at' => ['type' => 'DATETIME', 'null' => true],
            'deleted_at' => ['type' => 'DATETIME', 'null' => true],
        ])->addKey('id', true)->createTable('user', true); 
```

However if you wrote it like the following it would show up in the indexes:

```php
        $this->forge->addField([
            'id'         => ['type' => 'INTEGER', 'constraint' => 3],
            'name'       => ['type' => 'VARCHAR', 'constraint' => 80],
            'email'      => ['type' => 'VARCHAR', 'constraint' => 100],
            'country'    => ['type' => 'VARCHAR', 'constraint' => 40],
            'created_at' => ['type' => 'DATETIME', 'null' => true],
            'updated_at' => ['type' => 'DATETIME', 'null' => true],
            'deleted_at' => ['type' => 'DATETIME', 'null' => true],
        ])->addKey('id', true)->createTable('user', true); 
```
With this PR both cases will return indexes for primary key.

The need is to get a comprehensive list of database constraints. This will be useful in adding `upsert()` to the database.

Furthermore, this method doesn't use nested queries like the current implementation and requires only one database call.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by sclubricants** [comment]
Also, this adds the field index type - which all the other databases have. This tells whether the index is unique, primary, foreign, etc.


And something happened when rebasing my project.. There should  only be two files changed. Not sure what happend.

**Comment by sclubricants** [comment]
Any idea how all these other commits got here? Any idea how to fix?

**Comment by sclubricants** [comment]
Under the current implementation this:

```php
        $this->forge->addField([
            'id'         => ['type' => 'INTEGER', 'constraint' => 3, 'auto_increment' => true],
            'name'       => ['type' => 'VARCHAR', 'constraint' => 80],
            'email'      => ['type' => 'VARCHAR', 'constraint' => 100],
            'country'    => ['type' => 'VARCHAR', 'constraint' => 40],
            'created_at' => ['type' => 'DATETIME', 'null' => true],
            'updated_at' => ['type' => 'DATETIME', 'null' => true],
            'deleted_at' => ['type' => 'DATETIME', 'null' => true],
        ])->addKey(['id', 'email'], true)->createTable('user', true); 
```
will list an index for email but not for id. I know its an odd combination but interesting behavior.

**Comment by kenjis** [comment]
> Any idea how all these other commits got here? Any idea how to fix?

Rebase and force push.

```console
git switch SQLiteConnectiongetIndexData
git fetch upstream
git rebase upstream/4.3
git push --force-with-lease origin HEAD
```

File: system/Database/SQLite3/Connection.php [hunk 1]
```diff
@@ -265,9 +265,19 @@ protected function _fieldData(string $table): array
      */
     protected function _indexData(string $table): array
     {
-        // Get indexes
-        // Don't use PRAGMA index_list, so we can preserve index order
-        $sql = "SELECT name FROM sqlite_master WHERE type='index' AND tbl_name=" . $this->escape(strtolower($table));
+        $sql = "SELECT 'PRIMARY' as indexname, l.name as fieldname, 'PRIMARY' as indextype FROM pragma_table_info('" . trim($table, '`') . "') as l WHERE l.pk <> 0 UNION ALL
```
Comment (kenjis) [inline system/Database/SQLite3/Connection.php:268]: Too long. Please break the line.

Comment (kenjis) [inline system/Database/SQLite3/Connection.php:268]: Why do we need to ```trim($table, '`')```?

Comment (kenjis) [inline system/Database/SQLite3/Connection.php:268]: When we protect table name, we should use `protectIdentifiers()` method.

File: tests/system/Database/Live/SQLite/AlterTableTest.php [hunk 2]
```diff
@@ -277,4 +278,66 @@ protected function createTable(string $tableName = 'foo')
...
+        $this->forge->addField([
+            'id'         => ['type' => 'INTEGER', 'constraint' => 3, 'auto_increment' => true],
+            'userid'     => ['type' => 'INTEGER', 'constraint' => 3],
+            'name'       => ['type' => 'VARCHAR', 'constraint' => 80],
+            'email'      => ['type' => 'VARCHAR', 'constraint' => 100],
+            'country'    => ['type' => 'VARCHAR', 'constraint' => 40],
+            'created_at' => ['type' => 'DATETIME', 'null' => true],
+            'updated_at' => ['type' => 'DATETIME', 'null' => true],
+            'deleted_at' => ['type' => 'DATETIME', 'null' => true],
+        ])->addKey(['id'], true)->addUniqueKey('email')->addKey('country')->addForeignKey('userid', 'userforeign', 'id')->addForeignKey('email', 'userforeign', 'email')->createTable('testuser', true);
```
Comment (kenjis) [inline tests/system/Database/Live/SQLite/AlterTableTest.php:307]: Too long. Please break the line.

**Comment by kenjis** [comment]
`ForgeTest::testCompositeKey()` should be updated.

File: tests/system/Database/Live/SQLite/AlterTableTest.php [hunk 3]
```diff
@@ -277,4 +278,66 @@ protected function createTable(string $tableName = 'foo')
         $this->forge->addForeignKey('key_id', $tableName . '_fk', 'id');
         $this->forge->createTable($tableName);
     }
+
+    public function testGetIndexData()
```
Comment (kenjis) [inline tests/system/Database/Live/SQLite/AlterTableTest.php:282]: This test method has nothing to do with ALTER TABLE.
Please create a new TestCase file.

File: system/Database/SQLite3/Connection.php (hunk 1 above)
Comment (sclubricants) [reply system/Database/SQLite3/Connection.php:268]: We need 'table' not `table`

File: system/Database/SQLite3/Connection.php [hunk 4]
```diff
@@ -130,8 +130,9 @@ protected function execute(string $sql)
                 : $this->connID->query($sql);
         } catch (ErrorException $e) {
             log_message('error', (string) $e);
+
             if ($this->DBDebug) {
-                throw $e;
+                throw new DatabaseException($e->getMessage(), $e->getCode(), $e);
```
Comment (kenjis) [inline system/Database/SQLite3/Connection.php:135]: This change has nothing to do with this PR.
Please drop this change.

**Comment by kenjis** [comment]
There are a few commits that has nothing to do with this PR like https://github.com/codeigniter4/CodeIgniter4/pull/6207/commits/be0e5812dc8e82bfc917ed85430983ab57a14ee0, https://github.com/codeigniter4/CodeIgniter4/pull/6207/commits/2c731b72888d3423d821269976925a6ecea352be.
Please drop these commits.

File: system/Database/SQLite3/Connection.php (hunk 1 above)
Comment (kenjis) [reply system/Database/SQLite3/Connection.php:268]: Try `$this->protectIdentifiers($table)`.

I don't know we need `strtolower($table)` or not.
Can you check?

Comment (MGatner) [reply system/Database/SQLite3/Connection.php:268]: Aren't the back-ticks desired here? I just did a bunch of work with tables in SQLite3 and used them everywhere... maybe this function is some special exception?

Comment (MGatner) [reply system/Database/SQLite3/Connection.php:268]: Ah or maybe this is getting re-escaped below so the back-ticks were doubling?

**Comment by MGatner** [comment]
@sclubricants Any chance this will also fix https://github.com/codeigniter4/CodeIgniter4/issues/6205 by handling the indexes differently?

**Comment by sclubricants** [comment]
> There are a few commits that has nothing to do with this PR like [be0e581](https://github.com/codeigniter4/CodeIgniter4/commit/be0e5812dc8e82bfc917ed85430983ab57a14ee0), [2c731b7](https://github.com/codeigniter4/CodeIgniter4/commit/2c731b72888d3423d821269976925a6ecea352be). Please drop these commits.

How do I drop commits?

**Comment by sclubricants** [comment]
> @sclubricants Any chance this will also fix #6205 by handling the indexes differently?

I don't think this would affect anything with that. This affects only INTEGER AUTO_INCREMENT PRIMARY KEY. ..and also adds in the type UNIQUE, PRIMARY

**Comment by sclubricants** [comment]
Its important to note that something like the following would likely cause an error:

```php
        $keys = $db->getIndexData('tablename');
        
        foreach ($keys as $key) {
            // if($key->name != 'PRIMARY')
            $forge->dropKey('tablename', $key->name);            
        }
```

SQLite's handling of things the way they do creates a pseudo index but its not an index for the purposes of calling database commands. If you check that the index isn't name PRIMARY then the rest should be ok. 

In the unlikely case such as:

```php
        $this->forge->addField([
            'id'         => ['type' => 'INTEGER', 'constraint' => 3, 'auto_increment' => true],
            'name'       => ['type' => 'VARCHAR', 'constraint' => 80],
            'email'      => ['type' => 'VARCHAR', 'constraint' => 100],
            'country'    => ['type' => 'VARCHAR', 'constraint' => 40],
            'created_at' => ['type' => 'DATETIME', 'null' => true],
            'updated_at' => ['type' => 'DATETIME', 'null' => true],
            'deleted_at' => ['type' => 'DATETIME', 'null' => true],
        ])->addKey(['id', 'email'], true)->createTable('user', true); 
```

The id column doesn't get an index but the email column does. The way I wrote it the primary key index will assume the name of the email index but include both fields and have a type PRIMARY

**Comment by kenjis** [comment]
> How do I drop commits?

`git rebase -i`
https://opensource.com/article/22/4/manage-git-commits-rebase-i-command

**Comment by sclubricants** [comment]
I tried
git fetch upstream
git checkout master
fatal: 'master' matched multiple (2) remote tracking branches

**Comment by sclubricants** [comment]
$ git remote -v
origin  https://github.com/sclubricants/CodeIgniter4.git (fetch)
origin  https://github.com/sclubricants/CodeIgniter4.git (push)
upstream        https://github.com/codeigniter4/CodeIgniter4.git (fetch)
upstream        https://github.com/codeigniter4/CodeIgniter4.git (push)

**Comment by kenjis** [comment]
It seems you deleted `SQLiteConnectiongetIndexData` branch in your repository.

How to reset your `develop`:
```
git fetch upstream
git checkout 4.3
git branch -D develop
git checkout --track origin/develop
```

**Comment by kenjis** [comment]
It seems you deleted `SQLiteConnectiongetIndexData` branch in your repository.

How to reset your `develop`:
```
git fetch upstream
git checkout 4.3
git branch -D develop
git checkout --track upstream/develop
```

**Pull request closed without merging**